zebra-state/src/service.rs

```diff
@@ -724,17 +727,15 @@ impl StateService {
             self.queued_non_finalized_blocks
                 .prune_by_height(finalized_tip_height);
 
-            self.sent_non_finalized_block_hashes
-                .prune_by_height(finalized_tip_height);
+            self.sent_blocks.prune_by_height(finalized_tip_height);
         }
 
         rsp_rx
     }
 
     /// Returns `true` if `hash` is a valid previous block hash for new non-finalized blocks.
     fn can_fork_chain_at(&self, hash: &block::Hash) -> bool {
-        self.sent_non_finalized_block_hashes.contains(hash)
-            || &self.read_service.db.finalized_tip_hash() == hash
+        self.sent_blocks.contains(hash) || &self.read_service.db.finalized_tip_hash() == hash
```

Comment:
We can't fork the chain at a finalized block height below the finalized tip height, but this data structure now contains those heights.

Please store the queued finalized block list separately, rather than re-using an existing data structure that other code depends on.

Reply:
I think this might be an existing bug, we also can't fork from blocks that are near the final checkpoint, but we store their UTXOs in this data structure.

Reply:
I'll split this change into a follow-up PR and check that the finalized block sender has closed in `can_fork_at()`.